### PR TITLE
Fix painless casting bug causing opensearch to crash

### LINKED ISSUE #6435: [BUG] Opensearch process crashes when running a painless script
**Describe the bug**
OpenSearch process dies when running a painless script where a `def` type value is explicitly returned where the context is `void` type with following error.
```
» fatal error in thread [opensearch[runTask-0][write][T#5]], exiting
»  java.lang.NoClassDefFoundError: V
»       at org.opensearch.painless.PainlessScript$Script.execute(def x=1;return x;:16)
»       at org.opensearch.action.update.UpdateHelper.executeScript(UpdateHelper.java:337)
»       at org.opensearch.action.update.UpdateHelper.prepareUpdateScriptRequest(UpdateHelper.java:277)
»       at org.opensearch.action.update.UpdateHelper.prepare(UpdateHelper.java:107)
»       at org.opensearch.action.update.UpdateHelper.prepare(UpdateHelper.java:88)
»       at org.opensearch.action.update.TransportUpdateAction.shardOperation(TransportUpdateAction.java:224)
»       at org.opensearch.action.update.TransportUpdateAction.shardOperation(TransportUpdateAction.java:217)
»       at org.opensearch.action.update.TransportUpdateAction.shardOperation(TransportUpdateAction.java:89)
»       at org.opensearch.action.support.single.instance.TransportInstanceSingleOperationAction$ShardTransportHandler$1.doRun(TransportInstanceSingleOperationAction.java:321)
»       at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:806)
»       at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52)
»       at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
»       at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
»       at java.base/java.lang.Thread.run(Thread.java:829)
»  Caused by: java.lang.ClassNotFoundException: V
»       at java.base/java.lang.ClassLoader.findClass(ClassLoader.java:719)
»       at org.opensearch.painless.Compiler$Loader.findClass(Compiler.java:115)
»       at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
»       at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
»       ... 14 more
```

**To Reproduce**
1. Index some documents
```
curl -X PUT "localhost:9200/hockey/_bulk?refresh&pretty" -H 'Content-Type: application/json' -d'
{"index":{"_id":1}}
{"first":"johnny","last":"gaudreau","goals":[9,27,1],"assists":[17,46,0],"gp":[26,82,1],"born":"1993/08/13"}
{"index":{"_id":2}}
{"first":"sean","last":"monohan","goals":[7,54,26],"assists":[11,26,13],"gp":[26,82,82],"born":"1994/10/12"}
{"index":{"_id":3}}
{"first":"jiri","last":"hudler","goals":[5,34,36],"assists":[11,62,42],"gp":[24,80,79],"born":"1984/01/04"}
{"index":{"_id":4}}
{"first":"micheal","last":"frolik","goals":[4,6,15],"assists":[8,23,15],"gp":[26,82,82],"born":"1988/02/17"}
{"index":{"_id":5}}
{"first":"sam","last":"bennett","goals":[5,0,0],"assists":[8,1,0],"gp":[26,1,0],"born":"1996/06/20"}
{"index":{"_id":6}}
{"first":"dennis","last":"wideman","goals":[0,26,15],"assists":[11,30,24],"gp":[26,81,82],"born":"1983/03/20"}
{"index":{"_id":7}}
{"first":"david","last":"jones","goals":[7,19,5],"assists":[3,17,4],"gp":[26,45,34],"born":"1984/08/10"}
{"index":{"_id":8}}
{"first":"tj","last":"brodie","goals":[2,14,7],"assists":[8,42,30],"gp":[26,82,82],"born":"1990/06/07"}
{"index":{"_id":39}}
{"first":"mark","last":"giordano","goals":[6,30,15],"assists":[3,30,24],"gp":[26,60,63],"born":"1983/10/03"}
{"index":{"_id":10}}
{"first":"mikael","last":"backlund","goals":[3,15,13],"assists":[6,24,18],"gp":[26,82,82],"born":"1989/03/17"}
{"index":{"_id":11}}
{"first":"joe","last":"colborne","goals":[3,18,13],"assists":[6,20,24],"gp":[26,67,82],"born":"1990/01/30"}
'
```
2. Using a simple painless script in update context to replicate the error
```
curl -X POST "localhost:9200/hockey/_update/3?pretty" -H 'Content-Type: application/json' -d'
{
  "script": {
    "source": "def x=1;return x;",
    "lang": "painless"
  }
}
'
```

**Expected behavior**
OpenSearch process should not die and run the script as expected.

**Plugins**
N/A

**Host/Environment (please complete the following information):**
 - OS: MacOS
 - Version 13.2.1

### PR DESCRIPTION
**Pull request by shiv0408**
#### Update
Throwing ClassCastException when trying to cast def to void.
Created a Unit Test to ensure that def to void casting throws ClassCastException. Removed the following mention Unit Test as it was just added to reproduce the issue. The current unit test should be enough to avoid any regression in future.

Created a unit test to reproduce the bug #6435 by creating a unit test. Fix for this is requires a small change in  `modules/lang-painless/src/main/java/org/opensearch/painless/AnalyzerCaster.java` to throw `ClassCastException` when we are trying to cast `def` to `void`. But to create this unit test, I had to make change supported scripts in `_scripts/painless/_execute` API by adding the update context. How can we keep this unit test in this code without modifying the supported scripts? One option is we can keep a setting which users can't modify and set it from the test, using that setting allow the update operation in supported scripts.

### Related Issues
Resolves #6435 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
> I had to make change supported scripts in _scripts/painless/_execute API by adding the update context

So this is a new feature? I don't know this code well, what does this do? What are the downsides? If it's new, you should make a PR adding the feature first, then another one that fixes the bug we're talking about on top, just to keep things clean.

**Comment by shiv0408** [comment]
> So this is a new feature?

This change would allow users to run painless scripts with update context. Although, it is not my intention to add this as a new feature. Rather than creating a new feature, I wanted to keep context which is only used in tests and not by customers.

> what does this do? What are the downsides?

This [API](https://opensearch.org/docs/latest/api-reference/script-apis/exec-script/) allows customers to execute painless scripts with are not stored. Providing an Update context script has no major use case here.

**Comment by reta** [comment]
@shiv0408 could you please clarify what is supposed outcome of this script (from the issue):

```
curl -X POST "localhost:9200/hockey/_update/3?pretty" -H 'Content-Type: application/json' -d'
{
  "script": {
    "source": "def x=1;return x;",
    "lang": "painless",
    "params": {
      "sold_cost": 26
    }
  }
}
```

- the  `sold_cost` is not used - why it is there?
- the `ctx._source` is not used - nothing to update?

**Comment by shiv0408** [comment]
@reta I was modifying some script to reproduce the issue. The purpose of this script is just to demonstrate the issue, so did not use `ctx._source` to update any doc and `param` is also not required, left it by mistake.

**Comment by shiv0408** [comment]
And I am calling this update context, but any void context like `ingest` or `reindex` can face same issue if we try to return def type in a void expecting context.

**Comment by reta** [comment]
> And I am calling this update context, but any void context like `ingest` or `reindex` can face same issue if we try to return def type in a void expecting context.

So it the context is void expected - why script should return anything? And what is happening with the results that script returns? (as I see here, we just drop it)

**Comment by shiv0408** [comment]
Ideally, script should not return anything. But, if we do return something it should just throw exception or handle it gracefully and not crash the OpenSearch process.

**Comment by shiv0408** [comment]
@reta @dblock I have added the code which would throw the error when trying to cast from def to void type. But the UT I added in `PainlessExecuteApiTests.java` also passes now, but do we want to keep this test going ahead? I have also added a UT to ensure that we throw exception when trying to return def type from void context.

File: modules/lang-painless/src/main/java/org/opensearch/painless/action/PainlessExecuteAction.java [hunk 1]
```diff
@@ -577,6 +579,13 @@ static Response innerShardOperation(Request request, ScriptService scriptService
                     double result = scoreScript.execute(null);
                     return new Response(result);
                 }, indexService);
+            } else if (scriptContext == UpdateScript.CONTEXT) {
```
Comment (reta) [inline modules/lang-painless/src/main/java/org/opensearch/painless/action/PainlessExecuteAction.java:582]: We don't need the `UpdateScript.CONTEXT` right?

Comment (shiv0408) [reply modules/lang-painless/src/main/java/org/opensearch/painless/action/PainlessExecuteAction.java:582]: We don't need it to fix the bug. I added it to reproduce the issue in a Unit Test.

**Comment by shiv0408** [comment]
These changes are not necessary to fix the bug. I added them to repro the issue in a UT. As I have added a separate UT in `FactoryTests` we can remove this test.

**Review by reta — APPROVED** [review]

**Comment by shiv0408** [comment]
Thanks for approval @reta. Added DCO on all commits.

**Comment by reta** [comment]
@dblock looks good to you?

**Comment by shiv0408** [comment]
@reta Can we try to get this merged before 2.9 branch cut?

**Comment by reta** [comment]
> @reta Can we try to get this merged before 2.9 branch cut?

@dblock had comments / concerns, we need his approval

File: modules/lang-painless/src/test/java/org/opensearch/painless/FactoryTests.java [hunk 2]
```diff
@@ -358,6 +358,11 @@ public void testVoidReturn() {
             () -> getEngine().compile("void_return_test", "1 + 1", VoidReturnTestScript.CONTEXT, Collections.emptyMap())
         );
         assertEquals(iae.getMessage(), "not a statement: result not used from addition operation [+]");
+        ClassCastException exception = expectScriptThrows(
```
Comment (dblock) [inline modules/lang-painless/src/test/java/org/opensearch/painless/FactoryTests.java:361]: Make this test separate?

Comment (shiv0408) [reply modules/lang-painless/src/test/java/org/opensearch/painless/FactoryTests.java:364]: Created a different set in new commit

**Comment by shiv0408** [comment]
@dblock I think it should be good to merge now.

**Pull request merged**